src/Database/LSMTree/ModelIO/Normal.hs

```diff
@@ -0,0 +1,268 @@
+{-# LANGUAGE ConstraintKinds          #-}
+{-# LANGUAGE RecordWildCards          #-}
+{-# LANGUAGE StandaloneDeriving       #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TupleSections            #-}
+
+-- TODO: remove once the API is implemented.
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+-- |
+--
+-- This module is intended to be imported qualified.
+--
+-- > import qualified Database.LSMTree.Normal as LSMT
+module Database.LSMTree.ModelIO.Normal (
+    -- * Temporary placeholder types
+    Model.SomeSerialisationConstraint (..)
+    -- * Utility types
+  , IOLike
+    -- * Sessions
+  , Session
+  , newSession
+  , closeSession
+    -- * Tables
+  , TableHandle
+  , TableConfig (..)
+  , new
+  , close
+    -- * Table querying and updates
+    -- ** Queries
+  , Model.Range (..)
+  , Model.LookupResult (..)
+  , lookups
+  , Model.RangeLookupResult (..)
+  , rangeLookup
+    -- ** Updates
+  , Model.Update (..)
+  , updates
+  , inserts
+  , deletes
+    -- ** Blobs
+  , Model.BlobRef
+  , retrieveBlobs
+    -- * Snapshots
+  , VerificationFailure
+  , SnapshotName
+  , snapshot
+  , open
+    -- * Multiple writable table handles
+  , duplicate
+  ) where
+
+import           Control.Concurrent.Class.MonadMVar
+import           Data.Kind (Type)
+import           Database.LSMTree.Common (IOLike, Session, closeSession,
+                     newSession)
+import qualified Database.LSMTree.Model.Normal as Model
+
+type SomeSerialisationConstraint = Model.SomeSerialisationConstraint
+
+{-------------------------------------------------------------------------------
+  Tables
+-------------------------------------------------------------------------------}
+
+-- | A handle to a table.
+--
+-- An LSMT table is an individual key value mapping with in-memory and on-disk
+-- parts. A table handle is the object/reference by which an in-use LSM table
+-- will be operated upon. In our API it identifies a single mutable instance of
+-- an LSM table. The multiple-handles feature allows for there to may be many
+-- such instances in use at once.
+type TableHandle :: (Type -> Type) -> Type -> Type -> Type -> Type
+data TableHandle m k v blob = TableHandle {
+    thSession :: !(Session m)
+  , thRef     :: !(MVar m (Model.Table k v blob))
+  -- TODO: add a bit if this tablehandle is read only. Snapshots are read-only, AFAIU.
```

Comment:
A snapshot is also used for restarting when the program is killed, so a table handle opened from a snapshot should also be writable. IIRC we don't need to mark table handles as read-only for now, unless in the future we find a reason why we would have to distinguish between read-only and read-write table handles

Reply:
So what is difference between `snapshot` and `duplicate`? The documentation doesn't give a clue.

Recall, my model implementation is based only on the docs! Did I miss something?

Reply:
Now better explained in PR #10.